proposals/field-keyword-nullability.md

```diff
@@ -0,0 +1,89 @@
+# `field` keyword nullability
+
+## Goals
+
+Ensure a reasonable level of null-safety for various usage patterns of the `field` keyword feature. Avoid requiring the user to apply `[field: AllowNull, MaybeNull]` or similar attributes to the property in order to express common patterns around lazy initialization, etc.
+
+## Terms
+
+A property is a *field-backed property* if it meets any of the following conditions:
+1. It contains any automatically implemented accessors (`get;`/`set;`/`init;`).
+2. It uses the `field` keyword within any accessor bodies.
+
+Some examples of *field-backed properties* include:
+```cs
+public string Prop1 { get; set; }
+public string Prop2 { get => field; set; }
+public string? Prop3 { get => field; }
+```
+
+The variable denoted by the `field` keyword in a property's accessors is the *backing field* of that property.
+
+## Nullability of the *backing field*
+
+The *backing field* has the same type as the property. However, its nullable annotation may differ from the property. To determine this nullable annotation, we introduce the concept of *null-resiliency*. *Null-resiliency* intuitively means that the property's `get` accessor behaves properly even when the field contains the `default` value for its type.
+
+A *field-backed property* is determined to be *null-resilient* or not by performing a special nullable analysis of its get accessor.
+- For the purposes of this analysis, `field` is temporarily assumed to have *annotated* nullability, e.g. `string?`. This causes `field` to have *maybe-null* or *maybe-default* initial state in the get accessor, depending on its type.
+- Then, if nullable analysis of the getter yields no nullable warnings, the property is *null-resilient*. Otherwise, it is not *null-resilient*.
```

Comment:
No nullable warnings anywhere or just nullable warnings on `return` or `=>`? Be kinda odd if this method determined that the backing field was not null resistent

```csharp
string Prop { 
  get { 
    object o = null; // warning
    return field ?? "";
  }
}
```

Reply:
One reason we didn't want to scope things down to "returned a non-null value at all returns" was the following scenario:

```cs
string Prop
{
    get
    {
        field.Verify();
        return field;
    }
}
```

Imagine we are doing the "null-resilience" pass. field starts in the maybe-null state, so we warn on the dereference in `field.Verify()`, and update field's state to not-null.

Then, the return conversion check says cool, you returned a not-null value. We conclude that the property is null-resilient. Thus we think the field maybe-null, because the user dereferenced it.

We considered some fancier multi-pass stuff to say that if *additional* warnings are introduced when field starts as maybe-null, then it is not null-resilient. But we backed off of it for now due to the relative complexity of nullable analysis (difficulty of concluding that if the set of warnings changed, it was really field's nullability at fault.)

Reply:
@jaredpar Notes from our working group meeting about this: <https://github.com/dotnet/csharplang/blob/main/meetings/working-groups/field-keyword/FK-2024-08-07.md#open-question-unrelated-nullability-warnings>

Reply:
> difficulty of concluding that if the set of warnings changed, it was really field's nullability at fault

I feel very comfortable saying that _any other_ definition of fault would _not_ be a definition that we would want to use to determine null resilience. Given that the only difference between the two passes is the initial nullability of `field`, `field` is the key, and the user has not made it "safe" for `field` to be maybe-null. What else could we mean? It would be a good conversation to have.